### PR TITLE
Add test for D3D12 pixel history

### PR DESCRIPTION
**Pull request by skarolewics**
This mirrors the Vulkan pixel history test, but repeats the test for SM5.1, SM6.0, and SM6.6 (if available). Currently, the C++ backing for this test doesn't do any special binding for SM6.6, so the inclusion of that test is largely superficial for the time being.

### CODE REVIEW

**Review by Zorro666 — CHANGES_REQUESTED** [review]
Hey Steve

Thanks from local testing I found a couple of things that be good to fix up before merging the test

1. Some of the python is using the vulkan helpers and didn't run for me i.e. 
 - `post_mod_col` should be `get_post_mod_color`
 - `pre_mod_col` should be `get_pre_mod_color`
 - `post_mod_depth` should be `get_post_mod_depth`
2. I had to disable the MSAA test, it was failing for me `Expected 3 events, got 2` and when I look at the capture manually I think there should be three events but there are only two. I think there should be two primitives in the history but there is only one primitive.

Your call on what to do with the MSAA test. I would say disable it for this PR is fine then we have a known failure which can be worked on post-landing.

Be interesting to know if the test runs cleanly without errors for you locally. It might be a driver difference for the MSAA behaviour.

**Comment by skarolewics** [comment]
I addressed the issues you noticed in the test. The MSAA test succeeds for me, though I needed to make some changes to avoid crashes (another PR incoming for that). If the MSAA still fails with those fixes, can you share the output log from your test run?

**Review by Zorro666 — CHANGES_REQUESTED** [review]
Thanks Steve. The test still fails for me. Here is the Python log output

>> Test D3D12_Pixel_History
.. Running exe:'C:\Users\jake\Workspace\personal\renderdoc\util\test\demos_x64.exe' cmd:'D3D12_Pixel_History --log C:\Users\jake\Workspace\personal\renderdoc\util\test\tmp\demos.log' in dir:'' with env:'[]'
main.cpp:681 Log: Running 'D3D12_Pixel_History'
.. Queuing capture of frame 5..6 with timeout of default
d3d12\d3d12_test.cpp:255 Log: Running D3D12 test on NVIDIA GeForce RTX 4080 Laptop GPU
>> Raw Process output
d3d12\d3d12_test.cpp:255 Log: Running D3D12 test on NVIDIA GeForce RTX 4080 Laptop GPU
<< Raw Process output
.. Loading capture
.. Checking capture
.. Testing SM5.1
.. Testing primary
.. Testing pixel 190, 149
** Picked value at 190,149 in mainRT_SM51 is as expected
.. Testing pixel 190, 150
** Picked value at 190,150 in mainRT_SM51 is as expected
.. Testing pixel 200, 50
** Picked value at 200,50 in mainRT_SM51 is as expected
.. Testing pixel 150, 250
** Picked value at 150,250 in mainRT_SM51 is as expected
.. Testing pixel 330, 145
** Picked value at 330,145 in mainRT_SM51 is as expected
.. Testing pixel 340, 145
** Picked value at 340,145 in mainRT_SM51 is as expected
.. Testing pixel 330, 105
** Picked value at 330,105 in mainRT_SM51 is as expected
.. Testing pixel 320, 105
** Picked value at 320,105 in mainRT_SM51 is as expected
.. Testing pixel 345, 105
** Picked value at 345,105 in mainRT_SM51 is as expected
.. Testing dynamic state pipelines
.. Testing pixel 100, 250
** Picked value at 100,250 in mainRT_SM51 is as expected
.. Testing depth test for per fragment reporting
.. Testing pixel 275, 260
** Picked value at 275,260 in mainRT_SM51 is as expected
.. Testing secondary
.. Testing pixel 70, 40
** Picked value at 70,40 in mipArrayRT_SM51 is as expected
.. Testing pixel 40, 40
** Picked value at 40,40 in mipArrayRT_SM51 is as expected
** Picked value at 40,40 in mipArrayRT_SM51 is as expected
.. Testing depth
.. Testing pixel 200, 190
.. Testing MSAA
.. Testing pixel 140, 130 at sample 1
!+ FAILURE in D3D12_Pixel_History: TestFailureException Assertion Failure: Expected 3 events, got 2
>> Callstack
File "rdtest/testcase.py", line 205, in check
raise TestFailureException('Assertion Failure: {}'.format(msg))
File "tests/D3D12/D3D12_Pixel_History.py", line 423, in check_events
self.check(len(modifs) == len(events), "Expected {} events, got {}".format(len(events), len(modifs)))
    File "tests/D3D12/D3D12_Pixel_History.py", line 288, in multisampled_image_test
self.check_events(events, modifs, True)
    File "tests/D3D12/D3D12_Pixel_History.py", line 56, in check_capture
    self.multisampled_image_test("Begin " + t)
    File "rdtest/testcase.py", line 493, in run
self.check_capture()
    File "rdtest/testcase.py", line 500, in invoketest
self.run()
    File "rdtest/runner.py", line 349, in do
instance.invoketest(debugMode)
    File "rdtest/runner.py", line 357, in run_tests
do(False)
    << Callstack
    !- FAILURE
    $$ FAILED
    << Test D3D12_Pixel_History
    >> Raw RenderDoc log
    << Raw RenderDoc log
    // total=161 fail=1 skip=160 time=29

## Tests complete summary: 0 passed out of 1 run from 161 total in 0:00:29.579560 ##

    .. Failed tests:
    ..   - D3D12_Pixel_History

Here is what it looks like if I open the capture from the failed test and do the same pixel history manually.

Only one fragment is reported as touching the pixel but it should be two.
If I shader debug primitiveID zero and one then I get the expected colours from the shader debugger. If the Mesh Viewer I see both triangles. Looks like a bug in the pixel history code which for reason my driver exposes.

Happy to land this python file with the MSAA test disabled and then it can  be fixed up in a future PR. 

image

**Comment by Zorro666** [comment]
I did a bit of debugging of the MSAA issue and it looks like the compute shader dispatch to copy the stencil value from the MSAA Depth-Stencil buffer is reading zero back in the compute shader, but the actual buffer has the value two in it i.e. two fragments.

This leads to the pixel history thinking there is only one fragment touching the pixel

**Comment by skarolewics** [comment]
That MSAA copy shader was a pain to get working, I'm not entirely surprised that there's issues with it.  It passes with my GTX 970 but it could be a driver or GPU specific issue. I'll disable the MSAA test for now - you'll be much better able to identify the issue since you have the hardware exhibiting the issue.

**Review by Zorro666 — APPROVED** [review]
Thanks Steve - a great contribution and foundation to build on to get D3D12 Pixel History into RenderDoc

**Pull request merged**
